vhost-user-backend/src/bitmap.rs

```diff
@@ -0,0 +1,602 @@
+    // Any attempt to access beyond the end of the bitmap are simply ignored.
+    fn dirty_at(&self, offset: usize) -> bool {
+        let page = page_number(offset);
+        if page >= self.number_of_pages {
+            return false; // ignore out of bound access
+        }
+
+        // get the absolute page number
+        let page = self.pages_before_region + page;
+        let page_bit = self.logmem[page_word(page)].load(Ordering::Relaxed) & (1 << page_bit(page));
+        page_bit != 0
+    }
+}
+
+/// `MmaplogReg` mmaps the frontend bitmap backing memory in the current process.
+#[derive(Debug)]
+pub struct MmapLogReg {
+    addr: *const AtomicU8,
+    len: usize,
+}
+
+// SAFETY: Send is not automatically implemented because the raw pointer.
+// No one besides `MmapLogReg` has the raw pointer, so we can safely transfer it to another thread.
+unsafe impl Send for MmapLogReg {}
+
+// SAFETY: Sync is not automatically implemented because the raw pointer.
+// `MmapLogReg` doesn't have any interior mutability and all access to `&AtomicU8`
+// are done through atomic operations.
+unsafe impl Sync for MmapLogReg {}
+
+impl MmapLogReg {
+    #![allow(dead_code)]
+    // Note: We could try to adjust the mapping area to only cover the memory region, but
+    // the region's starting address is not guarantee to be LOG_WORD_SIZE-page aligned
+    // which makes the implementation needlessly cumbersome.
+    // Note: The specification does not define whether the offset must be page-aligned or not.
+    // But, since we are receiving the offset from the frontend to be used to call mmap,
+    // we assume it is properly aligned (currently, qemu always send a 0 offset).
+    pub(crate) fn from_file(fd: BorrowedFd, offset: u64, len: u64) -> io::Result<Self> {
+        let offset: isize = offset.io_try_into()?;
+        let len: usize = len.io_try_into()?;
+
+        // Let's uphold the safety contract for `std::ptr::offset()`.
+        if len > isize::MAX as usize {
+            return Err(io::Error::from(io::ErrorKind::InvalidData));
+        }
+
+        // SAFETY: `fd` is a valid file descriptor and we are not using `libc::MAP_FIXED`.
+        let addr = unsafe {
+            libc::mmap(
+                ptr::null_mut(),
+                len as libc::size_t,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_SHARED,
+                fd.as_raw_fd(),
+                offset as libc::off_t,
+            )
+        };
+
+        if addr == libc::MAP_FAILED {
+            return Err(io::Error::last_os_error());
+        }
+
+        Ok(Self {
+            addr: addr as *const AtomicU8,
+            len,
+        })
+    }
+
+    fn len(&self) -> usize {
+        self.len
+    }
+}
+
+impl Index<usize> for MmapLogReg {
+    type Output = AtomicU8;
+
+    // It's ok to get a reference to an atomic value.
+    fn index(&self, index: usize) -> &Self::Output {
+        assert!(index < self.len);
```

Comment:
can this be triggered by the frontend?

Reply:
I'm just thinking whether it's better to log an error and not handle the request, instead of crashing the device.

Reply:
sorry, I got confused (should not work when sick :) ), if hta assert fires is because a bug in `AtomicBitmapMmap`, what can I do is just ignore it, that is what `AtomicBitmapMmap` does if you try to set a bit outside the range of the bitmap

Reply:
> sorry, I got confused (should not work when sick :) ), that assert will fire only because a bug in `AtomicBitmapMmap`.

Okay, in that case I think it's fine.

> What I can do is just ignore it, that is what AtomicBitmapMmap does if you try to set a bit outside the range of the bitmap. But, as a second though, I don't want the "client" code to rely on that behavior (not panic) because it is what the Index::index will do for other types like Vec, and the idea (in the future) is to accept any type that implements Index.

If the panic is caused only by internal misbehavior, it's fine, but if it's triggerable from outside via UDS, then we should avoid it.